### PR TITLE
Add option allowing skip keys of nested object

### PR DESCRIPTION
**Pull request by petrnymsa**
Suppose we have the following translation JSON file

```json
{
    "generic": {
        "username": "Username"
    },
    "login": {
        "header" : {
            "title": "title",
            "someting": "sada"
        },
        "day": {
            "zero":"{} дней",
            "one": "{} день",
            "two": "{} дня",
            "few": "{} дня",
            "many": "{} дней",
            "other": "{} дней"
        }
    }
}
```

The current generator will generate the following keys structure, using `flutter pub run easy_localization:generate -S "assets/translations" -f keys -o locale_keys.g.dart`

```dart
abstract class  LocaleKeys {
  static const generic_username = 'generic.username';
  static const generic = 'generic';
  static const login_header_title = 'login.header.title';
  static const login_header_someting = 'login.header.someting';
  static const login_header = 'login.header';
  static const login_day = 'login.day';
  static const login = 'login';
}
````
Generated keys `generic`, `login` and `login_header` are not needed, as it will not provide any translation. 

This PR adds a new option to CLI enabling to **ignore** such not needed keys without breaking current behaviour for `plural()`, `gender()` functions. 

A new CLI flag is `--[no-]skip-object-keys`, abbreviation `x`.  

With this new flag, generated keys are following

```dart

abstract class  LocaleKeys {
  static const generic_username = 'generic.username';
  static const login_header_title = 'login.header.title';
  static const login_header_someting = 'login.header.someting';
  static const login_day = 'login.day';

}

```

As you can see, undesired keys are gone with preserving `login.day` for `plural()` function

I am open for any discussion, improving code quality.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -424,6 +424,46 @@ Code generation supports only json files, for more information run in terminal `
 | --output-dir | -O | lib/generated | Output folder stores for the generated file |
 | --output-file | -o | codegen_loader.g.dart | Output file name |
 | --format | -f | json | Support json or keys formats |
+| --[no-]skip-object-keys | -x | false | Ignores keys defining nested object except for pluarl(), gender() keywords. See example below  |
+
+`Skip-object-keys` example
```
Comment (Overman775) [inline README.md]: no example needed, can be confusing

File: README.md [hunk 2]
```diff
@@ -424,6 +424,46 @@ Code generation supports only json files, for more information run in terminal `
 | --output-dir | -O | lib/generated | Output folder stores for the generated file |
 | --output-file | -o | codegen_loader.g.dart | Output file name |
 | --format | -f | json | Support json or keys formats |
+| --[no-]skip-object-keys | -x | false | Ignores keys defining nested object except for pluarl(), gender() keywords. See example below  |
```
Comment (Overman775) [inline README.md]: `skip-unnecessary-keys` looking better

Comment (Overman775) [reply README.md]: then key need change to `-u`

File: bin/generate.dart [hunk 3]
```diff
@@ -73,6 +73,14 @@ ArgParser _generateArgParser(GenerateOptions? generateOptions) {
       help: 'Support json or keys formats',
       allowed: ['json', 'keys']);
 
+  parser.addFlag(
+    'skip-object-keys',
+    abbr: 'x',
+    defaultsTo: false,
+    callback: (bool? x) => generateOptions!.skipObjectKeys = x,
+    help: 'If true - skip keys of nested objects.',
```
Comment (Overman775) [inline bin/generate.dart]: Skip unnecessary keys of nested objects.

File: bin/generate.dart [hunk 4]
```diff
@@ -73,6 +73,14 @@ ArgParser _generateArgParser(GenerateOptions? generateOptions) {
       help: 'Support json or keys formats',
       allowed: ['json', 'keys']);
 
+  parser.addFlag(
+    'skip-object-keys',
+    abbr: 'x',
+    defaultsTo: false,
```
Comment (Overman775) [inline bin/generate.dart:79]: default true, its keys newer use, maybe ))

File: bin/generate.dart [hunk 5]
```diff
@@ -73,6 +73,14 @@ ArgParser _generateArgParser(GenerateOptions? generateOptions) {
       help: 'Support json or keys formats',
       allowed: ['json', 'keys']);
 
+  parser.addFlag(
+    'skip-object-keys',
+    abbr: 'x',
```
Comment (Overman775) [inline bin/generate.dart]: u

File: bin/generate.dart [hunk 6]
```diff
@@ -73,6 +73,14 @@ ArgParser _generateArgParser(GenerateOptions? generateOptions) {
       help: 'Support json or keys formats',
       allowed: ['json', 'keys']);
 
+  parser.addFlag(
+    'skip-object-keys',
```
Comment (Overman775) [inline bin/generate.dart]: skip-unnecessary-keys

File: bin/generate.dart [hunk 7]
```diff
@@ -83,10 +91,11 @@ class GenerateOptions {
   String? outputDir;
   String? outputFile;
   String? format;
+  bool? skipObjectKeys;
 
   @override
   String toString() {
-    return 'format: $format sourceDir: $sourceDir sourceFile: $sourceFile outputDir: $outputDir outputFile: $outputFile';
+    return 'format: $format sourceDir: $sourceDir sourceFile: $sourceFile outputDir: $outputDir outputFile: $outputFile skipObjectKeys: $skipObjectKeys';
```
Comment (Overman775) [inline bin/generate.dart]: skipUnnecessaryKeys

File: bin/generate.dart [hunk 8]
```diff
@@ -174,31 +183,45 @@ abstract class  LocaleKeys {
...
 }
 
-String _resolve(Map<String, dynamic> translations, [String? accKey]) {
+String _resolve(Map<String, dynamic> translations, bool? skipObjectKeys,
+    [String? accKey]) {
   var fileContent = '';
 
   final sortedKeys = translations.keys.toList();
 
+  final canIgnoreKeys = skipObjectKeys != null && skipObjectKeys;
```
Comment (Overman775) [inline bin/generate.dart]: bool get  canIgnoreKeys  => skipObjectKeys == true

**Review by Overman775 — CHANGES_REQUESTED** [review]

File: bin/generate.dart (hunk 4 above)
Comment (petrnymsa) [reply bin/generate.dart:79]: (sorry for delay and thanks for review). I've wanted to be by default false as it is "new feature" and would "break" current behavior..

File: bin/generate.dart (hunk 6 above)
Comment (petrnymsa) [reply bin/generate.dart]: Changed

File: bin/generate.dart (hunk 7 above)
Comment (petrnymsa) [reply bin/generate.dart]: Changed

File: README.md (hunk 1 above)
Comment (petrnymsa) [reply README.md]: Removed

File: README.md (hunk 2 above)
Comment (petrnymsa) [reply README.md]: Yeah sure, changed :]

File: bin/generate.dart (hunk 3 above)
Comment (petrnymsa) [reply bin/generate.dart]: Done

File: bin/generate.dart (hunk 4 above)
Comment (Overman775) [reply bin/generate.dart:79]: ok

File: bin/generate.dart [hunk 9]
```diff
@@ -161,8 +170,8 @@ void generateFile(
   printInfo('All done! File generated in ${outputPath.path}');
 }
 
-Future _writeKeys(
-    StringBuffer classBuilder, List<FileSystemEntity> files) async {
+Future _writeKeys(StringBuffer classBuilder, List<FileSystemEntity> files,
+    bool? skipObjectKeys) async {
```
Comment (Overman775) [inline bin/generate.dart]: skipUnnecessaryKeys

File: bin/generate.dart [hunk 10]
```diff
@@ -174,31 +183,45 @@ abstract class  LocaleKeys {
   Map<String, dynamic> translations =
       json.decode(await fileData.readAsString());
 
-  file += _resolve(translations);
+  file += _resolve(translations, skipObjectKeys);
```
Comment (Overman775) [inline bin/generate.dart]: skipUnnecessaryKeys

File: bin/generate.dart [hunk 11]
```diff
@@ -174,31 +183,45 @@ abstract class  LocaleKeys {
...
 
   classBuilder.writeln(file);
 }
 
-String _resolve(Map<String, dynamic> translations, [String? accKey]) {
+String _resolve(Map<String, dynamic> translations, bool? skipObjectKeys,
```
Comment (Overman775) [inline bin/generate.dart]: skipUnnecessaryKeys

File: bin/generate.dart [hunk 12]
```diff
@@ -174,31 +183,45 @@ abstract class  LocaleKeys {
...
+              translations[key] as Map<String, dynamic>) &&
+          canIgnoreKeys;
+
       var nextAccKey = key;
       if (accKey != null) {
         nextAccKey = '$accKey.$key';
       }
 
-      fileContent += _resolve(translations[key], nextAccKey);
+      fileContent += _resolve(translations[key], skipObjectKeys, nextAccKey);
```
Comment (Overman775) [inline bin/generate.dart]: skipUnnecessaryKeys

**Review by Overman775 — CHANGES_REQUESTED** [review]

File: bin/generate.dart (hunk 9 above)
Comment (petrnymsa) [reply bin/generate.dart]: Missed that.. changed

**Review by Overman775 — APPROVED** [review]

**Review by aissat — APPROVED** [review]

**Comment by petrnymsa** [comment]
@Overman775  @aissat  Hi, when can we expect to land new release of develop? This PR was merged one year ago

**Comment by aissat** [comment]
> @Overman775 @aissat Hi, when can we expect to land new release of develop? This PR was merged one year ago

Hello,
The PR has been merged, and landed the release.

**Pull request merged**